### DISCUSSION TITLE
modify + unwrap: Fix objects must not be overlapped in a report

### DISCUSSION THREAD

**Discussion by milahu** [eslint/eslint#16486] [Q&A] [answered]
im using eslint to build a [react2solid converter](https://github.com/milahu/eslint-plugin-react2solid)

i want to replace this react code

```jsx
export const Button = React.forwardRef(
  function Button(props, ref) {
    return <div ref={ref}>button</div>
  }
)
```

with this solid code

```jsx
export const Button =
  function Button(props) {
    return <div ref={props.ref}>button</div>
  }
```

so i need 2 overlapping fixes

1. modify the arg0 node (here: `function Button`)
   - remove the `ref` argument
   - replace the `ref` reference with `props.ref`
2. unwrap the arg0 node = remove `React.forwardRef( ... )`

when i do "modify" and "unwrap" in one `context.report` then eslint throws

```
AssertionError [ERR_ASSERTION]: Fix objects must not be overlapped in a report.
```

> in one `context.report`

i also tried 2 reports, but then only "unwrap" works

what did not work

see also [lib/rules/react-forward-ref-to-solid.js](https://github.com/milahu/eslint-plugin-react2solid/blob/ca63a705d3a4433b157de62b18668f4440dab918/lib/rules/react-forward-ref-to-solid.js)

```js

// unwrap the arg0 node

/**/
// AssertionError [ERR_ASSERTION]: Fix has invalid range
yield fixer.removeRange(node.range[0], arg0.range[0])
yield fixer.removeRange(arg0.range[1], node.range[1])
/**/

/**/
// no effect
for (const token of sourceCode.getTokensBetween(node, arg0)) {
  yield fixer.remove(token)
}
/**/

/**/
// debug: ok
// { node_range: [ 93, 275 ], arg0_range: [ 150, 273 ] }
// { node_range: [ 68, 241 ], arg0_range: [ 125, 239 ] }
console.dir({
  node_range: node.range,
  arg0_range: arg0.range,
})
/**/

/**/
// AssertionError [ERR_ASSERTION]: Fix has invalid range
yield fixer.removeRange(node.range[0]+1, arg0.range[0]-1)
/**/

/**/
// AssertionError [ERR_ASSERTION]: Fix objects must not be overlapped in a report.
yield fixer.replaceText(node, sourceCode.getText(arg0))
/**/

/**/
// debug: empty: { between1: [], between2: [] }
console.dir({
  between1: sourceCode.getTokensBetween(node, arg0),
  between2: sourceCode.getTokensBetween(arg0, node),
})
/**/

/**/
// no effect
fixer.removeRange(node.range[0], node.range[0]+57)
/**/
```

my last resort would be something like

```js
const oldCode = sourceCode.getText(arg0)
const ast = parse(oldCode)
// modify ast ...
const newCode = print(ast)
yield fixer.replaceText(node, newCode)
```

but i hope there is a better way

**Comment by milahu** [comment]
solved in [replaceReactForwardRef.ts](https://github.com/swordev/suid/blob/main/packages/codemod/src/transforms/replaceReactForwardRef.ts) of

[suid/codemod](https://github.com/swordev/suid/tree/main/packages/codemod) - Transform your MUI React code to SUID SolidJS. ([live demo](https://suid.io/tools/react-to-solid))

**Reply by milahu** [reply]
on closer look, this is also not possible with ts-morph

sure, i can do the naive approach

```js
call.replaceWithText(arg.getText());
```

but this breaks recursion (like in eslint)
so that the AST nodes in `arg` are not transformed

[suid/codemod](https://github.com/swordev/suid/tree/main/packages/codemod) solves this in [replaceReactForwardRef.ts](https://github.com/swordev/suid/blob/main/packages/codemod/src/transforms/replaceReactForwardRef.ts)

```js
const emptyMagicKey = `__EMPTY__${Date.now()}`;

const identifier = ...
const call = identifier.getFirstAncestorByKindOrThrow(ts.SyntaxKind.CallExpression);
const expr = call.getExpression();

addReplacePattern(identifier.getSourceFile(), emptyMagicKey, "");
expr.replaceWithText(emptyMagicKey);
call.getTypeArguments().map(arg => call.removeTypeArgument(arg));
```

where `expr` is the CallExpression#expression

```js
const someResult = someFunction(someArgument)
//                 ^^^^^^^^^^^^
//                 expr
```

so ```expr.replaceWithText(emptyMagicKey)``` gives


```js
const someResult = __EMPTY__1667548326458(someArgument)
```

and later ```jsaddReplacePattern(node.getSourceFile(), emptyMagicKey, "")```
will remove the `__EMPTY__1667548326458` string, leaving

```js
const someResult = (someArgument)
```

> addReplacePattern

this is actually part of [suid/codemod](https://github.com/swordev/suid/tree/main/packages/codemod/src/utils/addReplacePattern.ts)

in eslint, this should be possible too

**Comment by mdjermanovic** [comment, marked as answer]
`fixer.removeRange` expects 1 argument that is a 2-element array ([startIndex, endIndex]). You are passing in two arguments. Perhaps that's what's causing the problems in your code.

```diff
-yield fixer.removeRange(node.range[0], arg0.range[0])
+yield fixer.removeRange([node.range[0], arg0.range[0]])
```

https://github.com/eslint/eslint/blob/c97c7897686ac4dc2828537d6a017f3c99f7d905/lib/linter/rule-fixer.js#L123-L137

**Reply by milahu** [reply]
thanks!!

the actual fix was to enable type checking ...

jsconfig.json

```json
{
  "compilerOptions": {
    "checkJs": true
  },
  "exclude": [
    "node_modules",
    "**/node_modules/*"
  ]
}
```

which gives

```
Expected 1 arguments, but got 2. ts(2554)
```
